community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointScheduler.java

```diff
@@ -19,22 +19,32 @@
  */
 package org.neo4j.kernel.impl.transaction.log.checkpoint;
 
-import java.io.IOException;
+import java.util.Arrays;
 import java.util.function.BooleanSupplier;
 
 import org.neo4j.function.Predicates;
 import org.neo4j.kernel.impl.store.UnderlyingStorageException;
 import org.neo4j.kernel.impl.util.JobScheduler;
+import org.neo4j.kernel.internal.DatabaseHealth;
 import org.neo4j.kernel.lifecycle.LifecycleAdapter;
 
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 import static org.neo4j.kernel.impl.util.JobScheduler.Groups.checkPoint;
 
 public class CheckPointScheduler extends LifecycleAdapter
 {
+    /**
+     * The max number of consecutive check point failures that can be tolerated before treating
+     * check point failures more seriously, with a panic.
+     */
+    static final int MAX_CONSECUTIVE_FAILURES_TOLERANCE = 3;
+
     private final CheckPointer checkPointer;
     private final JobScheduler scheduler;
     private final long recurringPeriodMillis;
+    private final DatabaseHealth health;
+    private volatile Throwable[] failures = new Throwable[MAX_CONSECUTIVE_FAILURES_TOLERANCE];
```

Comment:
You said you were going to ensure that the contents of this array got cleared when it was no longer needed, I can't see that this is being done. This could be a reason for actually having this field `volatile` (probably the only good reason to do so): in order to be able to clear it by re-assigning a new array.